src/java.base/share/classes/com/sun/crypto/provider/AES_Crypt.java

```diff
@@ -0,0 +1,1418 @@
+            0x4697A38D,
+    };
+
+    /**
+     * Return the block cipher size.
+     *
+     * @return the AES block cipher size (in bytes).
+     */
+    int getBlockSize() {
+        return AESConstants.AES_BLOCK_SIZE;
+    }
+
+    /**
+     * Verifies if the length argument is a valid key size or not.
+     *
+     * @param len the size of the key (in bytes) to be validated.
+     *
+     * @return {@code true} if the size of the key is valid else {@code false}.
+     */
+    static boolean isKeySizeValid(int len) {
+        for (int kLength : AESConstants.AES_KEYSIZES) {
+            if (len == kLength) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check algorithm and initalize round keys for decryption and encryption.
+     *
+     * @param decrypting [in] indicates if encrypting ({@code false}) or
+     *      decrypting ({@code true}).
+     * @param algorithm [in] the case insentive string name for the AES cipher.
+     * @param key [in] the symmetric key byte array for encryption/decryption.
+     *
+     * @throws InvalidKeyException if an incorrect algorithm is given or if
+     * an invalid key size is provided.
+     */
+    void init(boolean decrypting, String algorithm, byte[] key)
+            throws InvalidKeyException {
+        int nk;
+        int decrypt = decrypting ? 1 : 0;
+
+        if (!algorithm.equalsIgnoreCase("AES")
+                && !algorithm.equalsIgnoreCase("Rijndael")) {
+            throw new InvalidKeyException ("Invalid algorithm name.");
+        }
+        if (key.length == AESConstants.AES_KEYSIZES[0]) {
+            rounds = AES_128_ROUNDS;
+            nk = AESConstants.AES_KEYSIZES[0]/WB;
+        } else if (key.length == AESConstants.AES_KEYSIZES[1]) {
+            rounds = AES_192_ROUNDS;
+            nk = AESConstants.AES_KEYSIZES[1]/WB;
+        } else if (key.length == AESConstants.AES_KEYSIZES[2]) {
+            rounds = AES_256_ROUNDS;
+            nk = AESConstants.AES_KEYSIZES[2]/WB;
+        } else {
+            throw new InvalidKeyException("Invalid key length (" + key.length
+                    + ").");
+        }
+        if (!MessageDigest.isEqual(prevKey, key)) {
+            if (sessionK == null) {
+                sessionK = new int[2][];
+            } else {
+                Arrays.fill(sessionK[0], 0);
+                Arrays.fill(sessionK[1], 0);
+            }
+            sessionK[0] = genRKeys(key, nk);
+            sessionK[1] = invGenRKeys();
+            if (prevKey != null) {
+                Arrays.fill(prevKey, (byte) 0);
+            }
+            prevKey = key.clone();
+        }
+        K = sessionK[decrypt];
+    }
+
+    /**
+     * Generate the cipher round keys.
+     *
+     * @param key [in] the symmetric key byte array.
+     * @param nk [in] the number of words in the key.
+     *
+     * @return w the cipher round keys.
+     */
+    private int[] genRKeys(byte[] key, int nk) {
```

Comment:
nit: The spec names this "keyExpansion" and documents it under section 5.2. Could we include this in the method javadoc? Also, how about "genRoundKeys" or "doKeyExpansion" as method name? `nk` can be derived from `key` and maybe no need for extra argument?